File: java/tests/com/facebook/yoga/utils/TestUtils.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.yoga.utils;

import com.facebook.yoga.YogaMeasureMode;
import com.facebook.yoga.YogaMeasureOutput;
import com.facebook.yoga.YogaMeasureFunction;
import com.facebook.yoga.YogaNode;
import com.facebook.yoga.YogaFlexDirection;

public class TestUtils {

  public static class intrinsicMeasureFunction implements YogaMeasureFunction {

    private static float widthPerChar = 10;
    private static float heightPerChar = 10;

    @Override
    public long measure(
        YogaNode node,
        float width,
        YogaMeasureMode widthMode,
        float height,
        YogaMeasureMode heightMode) {
      String text = (String) node.getData();
      float measuredWidth;
      float measuredHeight;

      if (widthMode == YogaMeasureMode.EXACTLY) {
        measuredWidth = width;
      } else if (widthMode == YogaMeasureMode.AT_MOST) {
        measuredWidth = Math.min(text.length() * widthPerChar, width);
      } else {
        measuredWidth = text.length() * widthPerChar;
      }

      if (heightMode == YogaMeasureMode.EXACTLY) {
        measuredHeight = height;
      } else if (heightMode == YogaMeasureMode.AT_MOST) {
        measuredHeight =
            Math.min(
                caclulateHeight(text, node.getFlexDirection() == YogaFlexDirection.COLUMN 
                  ? measuredWidth 
                  : Math.max(measuredWidth, getWidestWordWidth(text))), height);
      } else {
        measuredHeight = caclulateHeight(text, node.getFlexDirection() == YogaFlexDirection.COLUMN 
          ? measuredWidth 
          : Math.max(measuredWidth, getWidestWordWidth(text)));
      }

      return YogaMeasureOutput.make(measuredWidth, measuredHeight);
    }

    static float getWidestWordWidth(String text) {
      int widestWordLength = 0;
      String[] words = text.split(" ");
      for (String word : words) {
        int wordLength = word.length();
        if (widestWordLength < wordLength) {
          widestWordLength = wordLength;
        }
      }
      return (float) widestWordLength * widthPerChar;
    }

    static float caclulateHeight(String text, float measuredWidth) {
      if (text.length() * widthPerChar <= measuredWidth) {
        return heightPerChar;
      }

      String[] words = text.split(" ");
      float lines = 1;
      float currentLineLength = 0;
      for (String word : words) {
        float wordWidth = word.length() * widthPerChar;
        if (wordWidth > measuredWidth) {
          if (currentLineLength > 0) {
            lines++;
          }
          lines++;
          currentLineLength = 0;
        } else if (currentLineLength + wordWidth <= measuredWidth) {
          currentLineLength += wordWidth + widthPerChar;
        } else {
          lines++;
          currentLineLength = wordWidth + widthPerChar;
        }
      }
      return (currentLineLength == 0 ? lines - 1 : lines) * heightPerChar;
    }
  }
}

File: java/tests/com/facebook/yoga/utils/TestUtils.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.yoga.utils

import com.facebook.yoga.YogaFlexDirection
import com.facebook.yoga.YogaMeasureFunction
import com.facebook.yoga.YogaMeasureMode
import com.facebook.yoga.YogaMeasureOutput
import com.facebook.yoga.YogaNode

class TestUtils {
  class intrinsicMeasureFunction : YogaMeasureFunction {
    override fun measure(
        node: YogaNode,
        width: Float,
        widthMode: YogaMeasureMode,
        height: Float,
        heightMode: YogaMeasureMode,
    ): Long {
      val text = node.data as String
      val measuredWidth: Float =
          if (widthMode == YogaMeasureMode.EXACTLY) {
            width
          } else if (widthMode == YogaMeasureMode.AT_MOST) {
            Math.min(text.length * widthPerChar, width)
          } else {
            text.length * widthPerChar
          }

      val effectiveWidth =
          if (node.flexDirection == YogaFlexDirection.COLUMN) {
            measuredWidth
          } else {
            Math.max(measuredWidth, getWidestWordWidth(text))
          }

      val measuredHeight: Float =
          if (heightMode == YogaMeasureMode.EXACTLY) {
            height
          } else if (heightMode == YogaMeasureMode.AT_MOST) {
            Math.min(caclulateHeight(text, effectiveWidth), height)
          } else {
            caclulateHeight(text, effectiveWidth)
          }

      return YogaMeasureOutput.make(measuredWidth, measuredHeight)
    }

    companion object {
      private const val widthPerChar: Float = 10f
      private const val heightPerChar: Float = 10f

      @JvmStatic
      fun getWidestWordWidth(text: String): Float {
        var widestWordLength = 0
        val words = text.split(" ")
        for (word in words) {
          val wordLength = word.length
          if (widestWordLength < wordLength) {
            widestWordLength = wordLength
          }
        }
        return widestWordLength.toFloat() * widthPerChar
      }

      @JvmStatic
      fun caclulateHeight(text: String, measuredWidth: Float): Float {
        if (text.length * widthPerChar <= measuredWidth) {
          return heightPerChar
        }

        val words = text.split(" ")
        var lines = 1f
        var currentLineLength = 0f
        for (word in words) {
          val wordWidth = word.length * widthPerChar
          if (wordWidth > measuredWidth) {
            if (currentLineLength > 0) {
              lines++
            }
            lines++
            currentLineLength = 0f
          } else if (currentLineLength + wordWidth <= measuredWidth) {
            currentLineLength += wordWidth + widthPerChar
          } else {
            lines++
            currentLineLength = wordWidth + widthPerChar
          }
        }
        return (if (currentLineLength == 0f) lines - 1 else lines) * heightPerChar
      }
    }
  }
}
